flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -358,20 +358,27 @@ abstract class TableEnvironment(val config: TableConfig) {
     * Registers an [[AggregateFunction]] under a unique name. Replaces already existing
     * user-defined functions under this name.
     */
-  private[flink] def registerAggregateFunctionInternal[T: TypeInformation, ACC](
+  private[flink] def registerAggregateFunctionInternal[T: TypeInformation, ACC: TypeInformation](
       name: String, function: AggregateFunction[T, ACC]): Unit = {
     // check if class not Scala object
     checkNotSingleton(function.getClass)
     // check if class could be instantiated
     checkForInstantiation(function.getClass)
 
-    val typeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val resultTypeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val accTypeInfo: TypeInformation[_] = implicitly[TypeInformation[ACC]]
 
     // register in Table API
     functionCatalog.registerFunction(name, function.getClass)
 
     // register in SQL API
-    val sqlFunctions = createAggregateSqlFunction(name, function, typeInfo, typeFactory)
+    val sqlFunctions = createAggregateSqlFunction(
```

Comment:
To be honest, I don't think it is a nice approach to add the acc type to the internal representation of UDAGGs. It is not related to the logic of the function or required for the optimization but rather an internal runtime aspect.

An alternative could be to wrap the functions registered from Scala in a class that extends the `AggregateFunction` interface and holds the original agg function and the acc type information (exposed via the getAccumulatorType()` method). We would need to unwrap it before we translate it. It's not a super nice solution either, but would probably require fewer changes.

What do you think?